stylus-core/src/calls/context.rs

```diff
@@ -53,93 +36,27 @@ impl<S, const HAS_VALUE: bool> CallContext for Call<S, HAS_VALUE> {
     }
 }
 
-// allow &self as a context
-impl<T> CallContext for &T
-where
-    T: TopLevelStorage,
-{
-    fn gas(&self) -> u64 {
-        u64::MAX
-    }
-}
-
-// allow &mut self as a context
-impl<T> CallContext for &mut T
-where
-    T: TopLevelStorage,
-{
-    fn gas(&self) -> u64 {
-        u64::MAX
-    }
-}
-
-// allow &self to be a `pure` and `static` call context
-impl<T> StaticCallContext for &T where T: TopLevelStorage {}
+impl<S> StaticCallContext for Call<S, false> {}
 
-// allow &mut self to be a `pure` and `static` call context
-impl<T> StaticCallContext for &mut T where T: TopLevelStorage {}
+impl<S> NonPayableCallContext for Call<S, false> {}
 
-// allow &mut self to be a `write` and `payable` call context
-unsafe impl<T> MutatingCallContext for &mut T
-where
-    T: TopLevelStorage,
-{
+unsafe impl<S, const HAS_VALUE: bool> MutatingCallContext for Call<S, HAS_VALUE> {
```

Comment:
If we allow the dev to create the call like that, there isn't anything that prohibits a view function from calling a mutating one. For instance, the following code compiles:

```
sol_interface! {
    interface IFoo {
        function mutateFoo() external;
    }
}

#[entrypoint]
#[storage]
struct Contract {}

#[public]
impl Contract {
    pub fn view(&self, methods: IFoo) -> Result<(), Vec<u8>> {
        let cfg = Call::new();
        methods.mutate_foo(self.vm(), cfg).unwrap();
        Ok(())
    }
}
```

The MutatingCallContext trait should only be implemented for Call<&mut TopLevelTrait> to ensure a view function doesn't call a mutating one.